content/advent-2017/custom-JSON-unmarshaler-for-GraphQL-client.md

```diff
@@ -0,0 +1,558 @@
+
+[GraphQL](http://graphql.org/) is a data query language for APIs, developed internally by Facebook in 2012, and made publicly available in 2015. It can be used as a replacement for, or in addition to REST APIs. It some ways, it offers significant advantages compared to REST APIs, making it an attractive option. Of course, as any newer technology, it's less mature and has some weaknesses in certain areas.
+
+I spent a week on the initial investigation and research into what a Go client for GraphQL could look like. GraphQL is strongly typed, which is a good fit for Go. However, it also has some more advanced query syntax and features that play better with more dynamic languages, so I had my share of concerns whether a good client in Go would even be viable. Fortunately, at the end of that week, I found that a reasonable client in Go was indeed possible, and pushed a working [initial prototype](https://github.com/shurcooL/githubql/commit/78a7455460db3b5f51a2ec5640d7e47326a9ef12) that had most basic functionality implemented, with a plan for how to implement and address the [remaining features](https://github.com/shurcooL/githubql/issues/22).
+
+I documented the history of my findings and design decisions made in [this issue](https://github.com/google/go-github/issues/646). I've also given a [talk](https://www.youtube.com/watch?v=mEqJbeAazow) (slides [here](https://dmitri.shuralyov.com/talks/2017/githubql/githubql.slide)) about the rationale and thinking behind many of the API and design decisions that went into the library. In this post, I want to talk about something I haven't covered before: implementing a custom JSON unmarshaler specifically for the needs of the GraphQL client library, in order to improve support for GraphQL unions.
+
+JSON Unmarshaling Task at Hand
+------------------------------
+
+Unmarshaling JSON into a structure is a very common and well understood problem. It's already implemented inside the `encoding/json` package in Go standard library. Given that JSON is such a well specified standard, why would I need to implement my own JSON unmarshaler?
+
+To answer that, I need to provide a little context about how GraphQL works. The GraphQL client begins by sending a request containing a GraphQL query, for example:
+
+```GraphQL
+query {
+	me {
+		name
+		bio
+	}
+}
+```
+
+The GraphQL server receives it, processes it, and sends a JSON-encoded response for that query. The response contains a `data` object and potentially other miscellaneous fields. We're primarily interested in the `data` object, which looks like this:
+
+```JSON
+{
+	"me": {
+		"name": "gopher",
+		"bio": "The Go gopher."
+	}
+}
+```
+
+Notice it has the same shape as the query. Taking advantage of this property turned out to be a critical factor in making the Go GraphQL client library convenient and useful.
```

Comment:
I would delete the second sentence here and join the first one with the paragraph below.